### PR TITLE
refactor: remove char_u

### LINKED ISSUE #459: Remove char_u, long_u, short_u
We should replace most occurrences of  `char_u` by `uint8_t`(where only characters are used it's ok to use plain `char`) and use standard C99 fixed-size types such as `uint16_t`/`uint_64t` instead of `long_u`/`short_u`

# Wiki: https://github.com/neovim/neovim/wiki/Integer-types-refactoring-guidelines#char_u

### PR DESCRIPTION
**Pull request by dundargoc**
Closes https://github.com/neovim/neovim/issues/459

### CODE REVIEW

File: src/nvim/spell_defs.h [hunk 1]
```diff
@@ -243,7 +243,7 @@ typedef enum {
 
 typedef struct wordcount_S {
   uint16_t wc_count;                ///< nr of times word was seen
-  char_u wc_word[];                 ///< word
+  char wc_word[];                 ///< word
```
Comment (zeertzjq) [inline src/nvim/spell_defs.h:246]:
```suggestion
  char wc_word[];                   ///< word
```

File: src/nvim/types.h [hunk 2]
```diff
@@ -7,10 +7,6 @@
 // dummy to pass an ACL to a function
 typedef void *vim_acl_T;
 
-// Shorthand for unsigned variables. Many systems, but not all, have u_char
-// already defined, so we use char_u to avoid trouble.
-typedef unsigned char char_u;
```
Comment (clason) [inline src/nvim/types.h:12]: w00t 🥳

File: src/nvim/generators/gen_ex_cmds.lua [hunk 3]
```diff
@@ -41,7 +41,7 @@ static const uint16_t cmdidxs1[%u] = {
 -- Values in cmdidxs2[c1][c2] are relative to cmdidxs1[c1] so that they
 -- fit in a byte.
 local cmdidxs2_out = string.format([[
-static const char_u cmdidxs2[%u][%u] = {
+static const char cmdidxs2[%u][%u] = {
```
Comment (zeertzjq) [inline src/nvim/generators/gen_ex_cmds.lua:44]:
```suggestion
static const uint8_t cmdidxs2[%u][%u] = {
```

File: src/nvim/spell.c [hunk 4]
```diff
@@ -961,14 +961,14 @@ bool can_compound(slang_T *slang, const char *word, const uint8_t *flags)
 // compound rule.  This is used to stop trying a compound if the flags
 // collected so far can't possibly match any compound rule.
 // Caller must check that slang->sl_comprules is not NULL.
-bool match_compoundrule(slang_T *slang, const char_u *compflags)
+bool match_compoundrule(slang_T *slang, const char *compflags)
```
Comment (zeertzjq) [inline src/nvim/spell.c:964]:
```suggestion
bool match_compoundrule(slang_T *slang, const uint8_t *compflags)
```

File: src/nvim/spell.c [hunk 5]
```diff
@@ -150,7 +150,7 @@ typedef struct matchinf_S {
 
   // for when checking a compound word
   int mi_compoff;                       // start of following word offset
-  char_u mi_compflags[MAXWLEN];         // flags for compound words used
+  char mi_compflags[MAXWLEN];           // flags for compound words used
```
Comment (zeertzjq) [inline src/nvim/spell.c:153]:
```suggestion
  uint8_t mi_compflags[MAXWLEN];           // flags for compound words used
```

File: src/nvim/spell.c [hunk 6]
```diff
@@ -737,7 +737,7 @@ static void find_word(matchinf_T *mip, int mode)
         // If the word ends the sequence of compound flags of the
         // words must match with one of the COMPOUNDRULE items and
         // the number of syllables must not be too large.
-        mip->mi_compflags[mip->mi_complen] = (uint8_t)((unsigned)flags >> 24);
+        mip->mi_compflags[mip->mi_complen] = (char)(uint8_t)((unsigned)flags >> 24);
```
Comment (zeertzjq) [inline src/nvim/spell.c:740]:
```suggestion
        mip->mi_compflags[mip->mi_complen] = (uint8_t)((unsigned)flags >> 24);
```

File: src/nvim/spell.c [hunk 7]
```diff
@@ -750,7 +750,7 @@ static void find_word(matchinf_T *mip, int mode)
               xstrlcpy(fword, ptr, (size_t)endlen[endidxcnt] + 1);
             }
           }
-          if (!can_compound(slang, fword, mip->mi_compflags)) {
+          if (!can_compound(slang, fword, (uint8_t *)mip->mi_compflags)) {
```
Comment (zeertzjq) [inline src/nvim/spell.c:753]:
```suggestion
          if (!can_compound(slang, fword, mip->mi_compflags)) {
```

File: src/nvim/spellsuggest.c [hunk 8]
```diff
@@ -257,7 +257,7 @@ static bool can_be_compound(trystate_T *sp, slang_T *slang, uint8_t *compflags,
   if (slang->sl_comprules != NULL && sp->ts_complen > sp->ts_compsplit) {
     compflags[sp->ts_complen] = (uint8_t)flag;
     compflags[sp->ts_complen + 1] = NUL;
-    bool v = match_compoundrule(slang, compflags + sp->ts_compsplit);
+    bool v = match_compoundrule(slang, (char *)compflags + sp->ts_compsplit);
```
Comment (zeertzjq) [inline src/nvim/spellsuggest.c:260]:
```suggestion
    bool v = match_compoundrule(slang, compflags + sp->ts_compsplit);
```

File: src/nvim/spellfile.c [hunk 9]
```diff
@@ -5599,21 +5599,21 @@ void spell_add_word(char *word, int len, SpellAddType what, int idx, bool undo)
...
 
       // We just initialized the 'spellfile' option and can't open the
       // file.  We may need to create the "spell" directory first.  We
       // already checked the runtime directory is writable in
       // init_spellfile().
       if (!dir_of_file_exists(fname)
-          && (p = (char_u *)path_tail_with_sep(fname)) != (char_u *)fname) {
-        int c = *p;
+          && (p = path_tail_with_sep(fname)) != fname) {
+        int c = (uint8_t)(*p);
```
Comment (zeertzjq) [inline src/nvim/spellfile.c:5610]:
```suggestion
        char c = *p;
```

File: src/nvim/spellfile.c [hunk 10]
```diff
@@ -5599,21 +5599,21 @@ void spell_add_word(char *word, int len, SpellAddType what, int idx, bool undo)
...
 
         // The directory doesn't exist.  Try creating it and opening
         // the file again.
         *p = NUL;
         os_mkdir(fname, 0755);
-        *p = (char_u)c;
+        *p = (char)(uint8_t)c;
```
Comment (zeertzjq) [inline src/nvim/spellfile.c:5616]:
```suggestion
        *p = c;
```

File: src/nvim/spell.c [hunk 11]
```diff
@@ -961,14 +961,14 @@ bool can_compound(slang_T *slang, const char *word, const uint8_t *flags)
...
-bool match_compoundrule(slang_T *slang, const char_u *compflags)
+bool match_compoundrule(slang_T *slang, const uint8_t *compflags)
 {
   // loop over all the COMPOUNDRULE entries
   for (char *p = (char *)slang->sl_comprules; *p != NUL; p++) {
     // loop over the flags in the compound word we have made, match
     // them against the current rule entry
     for (int i = 0;; i++) {
-      int c = compflags[i];
+      int c = (uint8_t)compflags[i];
```
Comment (zeertzjq) [inline src/nvim/spell.c:971]:
```suggestion
      int c = compflags[i];
```

File: src/nvim/spell.c [hunk 12]
```diff
@@ -150,7 +150,7 @@ typedef struct matchinf_S {
 
   // for when checking a compound word
   int mi_compoff;                       // start of following word offset
-  char_u mi_compflags[MAXWLEN];         // flags for compound words used
+  uint8_t mi_compflags[MAXWLEN];           // flags for compound words used
```
Comment (zeertzjq) [inline src/nvim/spell.c:153]:
```suggestion
  uint8_t mi_compflags[MAXWLEN];        // flags for compound words used
```

**Review by zeertzjq — APPROVED** [review]

**Comment by zeertzjq** [comment]
Still draft?

**Pull request merged**